zcash_primitives/src/transaction/builder.rs

```diff
@@ -233,26 +242,26 @@ impl<'a, P: consensus::Parameters, R: RngCore> Builder<'a, P, R> {
     ///
     /// Returns an error if the given Merkle path does not have the same anchor as the
     /// paths for previous Sapling notes.
-    pub fn add_sapling_spend(
+    pub fn add_sapling_spend<FE>(
```

Comment:
I think that we should replace `FE` with `!` once that stabilizes.

Reply:
Hmm, that seems _worse_ from a UX perspective, because then the errors returned by this `Builder` API are incompatible with the errors returned by `Builder::build` and can't be handled together in any way. Returning `sapling::builder::Error` here would at least permit the caller to merge them if they want.

Reply:
I thought that `!` was going to be the bottom type? The bottom type should be treated by the compiler as every type, including the one you want here.

Reply:
The above example can be [fixed](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=4a18ed948da2de31e1525e20491fb048) by adding an `impl From<Error<!>> for Error<u32>`, but that can't be done generically as `impl<T> From<Error<!>> for Error<T>` because it collides with `impl<T> From<T> for T` (precisely because `!` can coerce to anything).